### PR TITLE
infoschema: add table_id to table and add tidb_indexes table.

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

now, we can query `table_id` and `index_id` from [http-api](https://github.com/pingcap/tidb/blob/master/docs/tidb_http_api.md), but in some enviorment, it'
s more convient to use sql.

final usage can be sawed in test case https://github.com/pingcap/tidb/pull/9183/files#diff-9cb2947f24dfdff2e6f6cec84b53aa56R212

### What is changed and how it works?

- add `tidb_table_id` column to `information_schema.tables` 
- add new `tidb_indexes` table that contains `index_id` column 

let use query `table_id` and `index_id` via sql.

### Check List 

Tests 

 - Unit test

Code changes

 - infomation schema: column and table

Side effects

 - N/A

Related changes

 - N/A

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/9183)

### CODE REVIEW

File: infoschema/tables.go [hunk 1]
```diff
@@ -995,6 +1010,7 @@ func dataForTables(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.D
 					nil,           // CHECKSUM
 					nil,           // CREATE_OPTIONS
 					"VIEW",        // TABLE_COMMENT
+					nil,           // TIDB_TABLE_ID
```
Comment (crazycs520) [inline infoschema/tables.go]: View also have table ID. But the column name is `TIDB_TABLE_ID`, return nil here is ok for me.

File: infoschema/tables.go [hunk 2]
```diff
@@ -1003,6 +1019,68 @@ func dataForTables(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.D
...
+						pkCol = col
+						break
+					}
+				}
+				record := types.MakeDatums(
+					schema.Name.O, // TABLE_SCHEMA
+					tb.Name.O,     // TABLE_NAME
+					0,             // NON_UNIQUE
+					"PRIMARY",     // KEY_NAME
+					1,             // SEQ_IN_INDEX
```
Comment (lamxTyler) [inline infoschema/tables.go]: Is 0 better? Or it may coincide　with indices?

File: infoschema/tables.go [hunk 3]
```diff
@@ -530,6 +532,18 @@ var tableProcesslistCols = []columnInfo{
...
 }
 
+var tableTiDBIndexesCols = []columnInfo{
+	{"TABLE_SCHEMA", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"TABLE_NAME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"NON_UNIQUE", mysql.TypeLonglong, 21, 0, nil, nil},
+	{"KEY_NAME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"SEQ_IN_INDEX", mysql.TypeLonglong, 21, 0, nil, nil},
+	{"COLUMN_NAME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"SUB_PART", mysql.TypeLonglong, 21, 0, nil, nil},
```
Comment (XuHuaiyu) [inline infoschema/tables.go]: Why not name it as `LENGTH`?

File: infoschema/tables.go [hunk 4]
```diff
@@ -1003,6 +1019,68 @@ func dataForTables(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.D
...
+				for _, col := range tb.Cols() {
+					if mysql.HasPriKeyFlag(col.Flag) {
+						pkCol = col
+						break
+					}
+				}
+				record := types.MakeDatums(
+					schema.Name.O, // TABLE_SCHEMA
+					tb.Name.O,     // TABLE_NAME
+					0,             // NON_UNIQUE
```
Comment (XuHuaiyu) [inline infoschema/tables.go]: May this 0 confuse others?

File: infoschema/tables.go [hunk 5]
```diff
@@ -530,6 +532,18 @@ var tableProcesslistCols = []columnInfo{
 	{"Info", mysql.TypeString, 512, 0, nil, nil},
 }
 
+var tableTiDBIndexesCols = []columnInfo{
+	{"TABLE_SCHEMA", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"TABLE_NAME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"NON_UNIQUE", mysql.TypeLonglong, 21, 0, nil, nil},
+	{"KEY_NAME", mysql.TypeVarchar, 64, 0, nil, nil},
```
Comment (XuHuaiyu) [inline infoschema/tables.go]: put this column before `NON_UNIQUE`?

Comment (lysu) [reply infoschema/tables.go]: it's following `show index`

File: infoschema/tables.go (hunk 2 above)
Comment (lysu) [reply infoschema/tables.go]: maybe we should better let seq start from 1 https://dev.mysql.com/doc/refman/8.0/en/show-index.html

File: infoschema/tables.go [hunk 6]
```diff
@@ -1003,6 +1019,68 @@ func dataForTables(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.D
 	return rows, nil
 }
 
+func dataForIndexes(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Datum, error) {
```
Comment (eurekaka) [inline infoschema/tables.go]: This looks like duplicate code with `fetchShowIndex`? can we extract it to a separate function for reuse?

File: infoschema/tables.go [hunk 7]
```diff
@@ -66,6 +66,7 @@ const (
 	tableTableSpaces                        = "TABLESPACES"
 	tableCollationCharacterSetApplicability = "COLLATION_CHARACTER_SET_APPLICABILITY"
 	tableProcesslist                        = "PROCESSLIST"
+	tableTiDBIndexes                        = "TIDB_INDEXES"
```
Comment (eurekaka) [inline infoschema/tables.go]: Directly use `INDICES` as table name?

File: infoschema/tables.go [hunk 8]
```diff
@@ -147,6 +148,7 @@ var tablesCols = []columnInfo{
 	{"CHECKSUM", mysql.TypeLonglong, 21, 0, nil, nil},
 	{"CREATE_OPTIONS", mysql.TypeVarchar, 255, 0, nil, nil},
 	{"TABLE_COMMENT", mysql.TypeVarchar, 2048, 0, nil, nil},
+	{"TIDB_TABLE_ID", mysql.TypeLonglong, 21, 0, nil, nil},
```
Comment (eurekaka) [inline infoschema/tables.go]: s/TIDB_TABLE_ID/TABLE_ID?

Comment (lysu) [reply infoschema/tables.go]: better not, it's special column not standard column

File: infoschema/tables.go (hunk 7 above)
Comment (lysu) [reply infoschema/tables.go]: better not, index is nore standard sql concept, for example: mysql use `innodb_indexes`

File: infoschema/tables.go (hunk 2 above)
Comment (lamxTyler) [reply infoschema/tables.go]: Ok, so it is better to make the `seq_in_index` be `len(rows)+1` for index.

File: infoschema/tables_test.go [hunk 9]
```diff
@@ -208,3 +208,19 @@ func (s *testSuite) TestViews(c *C) {
...
+	store, err := mockstore.NewMockTikvStore()
+	c.Assert(err, IsNil)
+	defer store.Close()
+	session.SetStatsLease(0)
+	do, err := session.BootstrapSession(store)
+	c.Assert(err, IsNil)
+	defer do.Close()
+	tk := testkit.NewTestKit(c, store)
+	tk.MustExec("create table test.t (a int, b int, primary key(a), key k1(b))")
+	tk.MustQuery("select tidb_table_id from information_schema.tables where table_schema = 'test' and table_name = 't'").Check(testkit.Rows("31"))
```
[the remark this answers is not in the document]
Comment (eurekaka) [reply infoschema/tables_test.go]: IMO it is better to just check the field name is `tidb_table_id` and the field value is larger than 0.

**Pull request merged**